[1.19.X] Client code cleanup, updates, and other refactors

amadornes opened the pull request:
As has been discussed both publicly and internally, several parts of the client codebase have been in need of maintenance in one way or another for some time now. The team has a lot of responsibilities and it's just not possible for them to dedicate all their time and energy for close to 15 hours a day over the span of 3 weeks to cleaning things up and getting them up to spec.

But as it turns out, I do lol

### This PR is the result of a larger effort with several goals:
 - Audit and document the purpose of all current client code. This provides better insight into what exactly we're dealing with
 - Clean up the codebase, removing everything that is not in use and updating older systems to match current standards and user demands. This removes tech debt and improves both maintainability and user-facing accessibility. Some examples:
     - Adapt all code to current naming conventions
     - Reorganize and rename classes and members to improve organization and legibility
     - Move away from static registration in `FMLClientSetupEvent` and over to dedicated registration events
     - Remove old vertex pipelines and substitute them with new ones based on vanilla's `VertexConsumer` interface
     - Decouple flat and shaded lighting pipelines since they are completely unrelated, and make them into vertex producers instead of part of the vertex pipeline
     - Improve functionality of vanilla's model display transforms, removing the need for Forge's largely overlapping replacement
 - Introduce several breaking or semi-breaking changes, like:
     - Make `IModelData` into a final class backed by an immutable map, solving thread safety concerns
     - Move model data manager ownership to `ClientLevel`, solving issues with multiple active client levels
     - Make render types a first-class citizen in `BakedModel`. This also involves transferring over control of render types from `ItemBlockRenderTypes`, and may result in a slight speedup in chunk geometry baking
 - Provide clear and concise documentation for the most relevant elements of the client API as well as an update guide (WIP)

The above listed changes are not an exhaustive list and more can be found in the comments of each commit.

### Non-goals of this PR:
 - Clean up hooks in patches. This is very much needed, as a large number of events and other hooks are dispatched directly from Minecraft classes instead of through `ForgeHooksClient`, but is outside of the scope of this PR as these would be non-breaking changes for developers and could therefore be safely introduced mid-version
 - Optimize existing code for speed and/or cleanliness. While minor improvements have been made in multiple areas, this is generally just a side effect of the updating/upgrading effort, and not a direct focus of this PR

### Code organization:
As per the request of the triage team, the PR has been squashed and consequently split into commits organized by "systems" or largely related units of code. These commits will not compile on their own, and it is likely they do not contain all the classes referenced in the code, as a lot of files have been touched by these changes and some of them belong in other commits. Each commit contains a general list of changes, and virtually all files in the client package have javadoc explaining their purpose.

### A few words
Thank you to everyone who helped make this happen, from the Forge core team members, to contributors, triage and third parties for their help and patience during the documentation and initial review phases of the audit and changelist, and even testing the changes with their mods.

I've reviewed everything twice over and all seems in order. Excluding a few minor tweaks made during implementation, the proposed changes have already been discussed with members of the team (this doesn't imply full approval, but it's likely not far from it), so at this point it's down to triage to spot any silly mistakes and oversights I've missed.  
Good luck with the 284 files. I'll be here and on Discord if you have any questions 😅 

Edit: Here is the original audit doc with the planned changes, if anyone is curious: https://forge-render-refactor.notion.site/forge-render-refactor/ecd69011d01042c48a0fca80696cb4da?v=47332327988948ebaf911b06b2aa1f5d

Edit 2: Fixes #8362

amadornes committed:
Renderable API refactors
- Rename "render values" to "context"
- Rename SimpleRenderable to CompositeRenderable to better reflect its use
- Remove IMultipartRenderValues since it doesn't have any real use
- Add extensive customization options to BakedModelRenderable

amadornes committed:
ClientRegistry and MinecraftForgeClient refactors
- Add sprite loader manager and registration event
- Add spectator shader manager and registration event
- Add client tooltip factory manager and registration event
- Add recipe book manager and registration event
- Add key mapping registration event
- Remove ClientRegistry, as everything has been moved out of it
- Remove registration methods from MinecraftForgeClient, as they have dedicated events now

amadornes committed:
Dimension special effects refactors
- Fold handlers into an extension class and remove public mutable fields
- Add dimension special effects manager and registration event

amadornes committed:
HUD overlay refactors
- Rename to IGuiOverlay match vanilla (instead of Ingame)
- Add overlay manager and registration event
- Move vanilla overlays to a standalone enum

amadornes committed:
Model loader refactors
- Rename IModelLoader to IGeometryLoader
- Add loader manager and registration event
- Fold all model events into one
- Move registration of additionally loaded models to an event
- Remove ForgeModelBakery and related classes as they served no purpose anymore

amadornes committed:
Render properties refactors
- Rename all render properties to client extensions and relocate accordingly
- Move lookups to the respective interfaces

amadornes committed:
Model data refactors
- Convert model data to a final class backed by an immutable map and document mutability requirements. This addresses several thread-safety issues in the current implementation which could result in race conditions
- Transfer ownership of the data manager to the client level. This addresses several issues that arise when multiple levels are used at once

amadornes committed:
GUI and widget refactors
- Move all widgets to the correct package
- Rename GuiUtils and children to match vanilla naming

amadornes committed:
New vertex pipeline API
- Move to vanilla's VertexConsumer
- Roll back recent PR making VertexConsumer format-aware. This is the opposite of what vanilla does, and should not be relevant with the updated lighting pipeline

amadornes committed:
Lighting pipeline refactors
- Move to dedicated lighting package
- Separate flat and smooth lighters
- Convert from a vertex pipeline transformer to a pure vertex source (input is baked quads)

amadornes committed:
Model geometry API refactors
- Rename IModelGeometry to IUnbakedGeometry
- Rename IModelConfiguration to IGeometryBakingContext
- Rename other elements to match vanilla naming
- Remove current changes to ModelState, as they do not belong there. Transforms should be specified through vanilla's system. ModelState is intended to transfer state from the blockstate JSON
- Remove multipart geometries and geometry parts. After some discussion, these should not be exposed. Instead, geometries should be baked with only the necessary parts enabled

amadornes committed:
Make render types a first-class citizen in baked models
- Add named render types (block + entity + fabulous entity)
- Add named render type manager + registration event
- Make BakedModel aware of render types and transfer control over which ones are used to it instead of ItemBlockRenderTypes (fallback)
- (additional) Add concatenated list view. A wrapper for multiple lists that iterates through them in order without the cost of merging them. Useful for merging lists of baked quads

amadornes committed:
General event refactors
- Several renames to either match vanilla or improve clarity
- Relocate client chat event dispatching out of common code

amadornes committed:
Forge model type refactors
- Rename SeparatePerspectiveModel to SeparateTransformsModel
- Rename ItemModelMesherForge to ForgeItemModelShaper
- Rename DynamicBucketModel to DynamicFluidContainerModel
- Prefix all OBJ-related classes with "Obj" and decouple parsing from construction
- Extract ElementsModel from model loader registry
- Add EmptyModel (baked, unbaked and loader)
- Refactor CompositeModel to take over ItemMultiLayerBakedModel
- Remove FluidModel as it's not used and isn't compatible with the new fluid rendering in modern versions
- Move model loader registration to a proper event handler
- Update names of several JSON fields (backwards-compatible)
- Update datagens to match

amadornes committed:
Update test mods

amadornes committed:
Miscellaneous changes and overlapping patches
- Dispatch all new registration events
- Convert ExtendedServerListData to a record
- Add/remove hooks from ForgeHooksClient as necessary

XFactHD commented on patches/minecraft/net/minecraft/client/color/block/BlockColors.java.patch:
```diff
@@ -33,6 +33,7 @@
        return blockcolor == null ? -1 : blockcolor.m_92566_(p_92578_, p_92579_, p_92580_, p_92581_);
     }
  
++   @Deprecated
```
Should have a comment explaining why it is deprecated

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on patches/minecraft/net/minecraft/client/color/item/ItemColors.java.patch:
```diff
@@ -24,6 +24,7 @@
        return itemcolor == null ? -1 : itemcolor.m_92671_(p_92677_, p_92678_);
     }
  
++   @Deprecated
```
Should have a comment explaining why it is deprecated

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on patches/minecraft/net/minecraft/client/particle/ParticleEngine.java.patch:
```diff
@@ -14,14 +14,17 @@
+       this.m_107378_(ParticleTypes.f_235901_, ShriekParticle.Provider::new);
     }
  
++   @Deprecated
```
Should have a comment explaining why it is deprecated

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on patches/minecraft/net/minecraft/client/particle/ParticleEngine.java.patch:
```diff
@@ -14,14 +14,17 @@
 +      this.f_107293_.put(Registry.f_122829_.m_7981_(p_107382_), p_107383_);
     }
  
++   @Deprecated
```
Should have a comment explaining why it is deprecated

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on patches/minecraft/net/minecraft/client/renderer/ItemBlockRenderTypes.java.patch:
```diff
@@ -38,64 +38,63 @@
-+      return fluidRenderChecks.get(net.minecraftforge.registries.ForgeRegistries.FLUIDS.getDelegateOrThrow(fluid.m_76152_())).test(type);
-+   }
-+
++   @Deprecated(forRemoval = true, since = "1.19")
```
Should have a comment explaining why it is deprecated

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on patches/minecraft/net/minecraft/client/renderer/ItemBlockRenderTypes.java.patch:
```diff
@@ -38,64 +38,63 @@
++      setRenderLayer(block, java.util.Set.of(type));
 +   }
 +
++   @Deprecated(forRemoval = true, since = "1.19")
```
Should have a comment explaining why it is deprecated

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on patches/minecraft/net/minecraft/client/renderer/ItemBlockRenderTypes.java.patch:
```diff
@@ -38,64 +38,63 @@
 +
-+   public static void setRenderLayer(Fluid fluid, RenderType type) {
-+      setRenderLayer(fluid, createMatchingLayerPredicate(type));
++   @Deprecated(forRemoval = true, since = "1.19")
```
Should have a comment explaining why it is deprecated

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on patches/minecraft/net/minecraft/client/resources/model/SimpleBakedModel.java.patch:
```diff
@@ -1,13 +1,73 @@
  
-+      public Builder(net.minecraftforge.client.model.IModelConfiguration model, ItemOverrides overrides) {
-+         this(model.useSmoothLighting(), model.isSideLit(), model.isShadedInGui(), model.getCameraTransforms(), overrides);
++   @Deprecated // Forge: Use variant with RenderTypeGroup
```
Deprecation comment should be a javadoc comment

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on patches/minecraft/net/minecraft/client/resources/model/SimpleBakedModel.java.patch:
```diff
@@ -1,13 +1,73 @@
+          return this;
+       }
+ 
++      @Deprecated // Forge: Use variant with RenderType
```
Deprecation comment should be a javadoc comment

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

amadornes committed:
Fix VertexConsumerWrapper returning parent instead of itself

XFactHD commented on src/main/java/net/minecraftforge/client/event/RegisterKeyMappingsEvent.java:
```diff
@@ -5,32 +5,37 @@
+     * Registers a new key mapping.
      */
-    public ModelRegistryEvent()
+    public synchronized void register(KeyMapping key)
```
The syncronisation is unnecessary, the event doesn't fire in paralell.

amadornes replied:
Good point. Seems I forgot to ask about parallel dispatch. The new registration events should likely fire in parallel, since they're not inter-dependent and are already being fired with concurrent collections.

amadornes replied:
Fixed in [7bdcf52](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/7bdcf52de92f0c8289bf5d84a83a4063ba5213e8)

XFactHD commented on src/main/java/net/minecraftforge/client/model/IModelBuilder.java:
```diff
@@ -6,50 +6,103 @@
         public BakedModel build()
         {
-            return builder.build();
+            return EmptyModel.BAKED;
```
This doesn't seem right, shouldn't this return a model with all the collected quads?

amadornes replied:
It collects quads to a user-provided list, but it lacks any of the context to actually build it into a model that the other built-in model builder has. I might want to document the factory method to say that it will always return an empty model, though.

XFactHD replied:
A doc comment would definitely help, this is a bit unintuitive as it stands.

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on src/main/java/net/minecraftforge/client/extensions/IForgeVertexConsumer.java:
```diff
@@ -106,7 +99,7 @@ default void putBulkData(PoseStack.Pose pose, BakedQuad bakedQuad, float[] baseB
         int bl = packedLight&0xFFFF;
         int sl = (packedLight>>16)&0xFFFF;
-        int offset = LightUtil.getLightOffset(0) * 4; // int offset for vertex 0 * 4 bytes per int
+        int offset = 6 * 4; // int offset for vertex 0 * 4 bytes per int
```
This should probably use the `IQuadTransformer.UV2` constant instead of the magic number `6`

amadornes replied:
Added in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on src/main/java/net/minecraftforge/client/model/data/ModelDataManager.java:
```diff
@@ -0,0 +1,101 @@
+
+/**
+ * A manager for the lifecycle of all the {@link ModelData} instances in a {@link Level}.<p/>
+ * Users should not be instantiating or using this themselves.
```
This is not true in my opinion, modders may very well have a good reason to instantiate this themselves, specifically for completely custom fake client worlds, i.e. how Create renders Contraptions.

amadornes replied:
I'd say it's fine to leave the comment in, since the only cases where people might want to instantiate it themselves are edge cases where they have to know what they're doing one way or another, and if they want/have to, they will - nobody can stop them. But if people agree it should be changed, I can change it

amadornes replied:
Changed in [848f3ed](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/848f3ed9383583289102ff4859d2cf48b23b5360)

XFactHD commented on src/main/java/net/minecraftforge/client/model/generators/loaders/ItemLayersModelBuilder.java:
```diff
@@ -23,17 +24,83 @@ public static <T extends ModelBuilder<T>> ItemLayersModelBuilder<T> begin(T pare
-        fullbright.add(layer);
+        Preconditions.checkNotNull(layers, "Layers must not be null");
+        Preconditions.checkArgument(layers.length > 0, "At least one layer must be specified");
+        Preconditions.checkArgument(Arrays.stream(layers).anyMatch(i -> i < 0), "All layers must be >= 0");
```
This should either be `allMatch(i -> i >= 0)` or `noneMatch(i -> i < 0)`

amadornes replied:
Fixed in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on src/main/java/net/minecraftforge/client/model/generators/loaders/ItemLayersModelBuilder.java:
```diff
@@ -23,17 +24,83 @@ public static <T extends ModelBuilder<T>> ItemLayersModelBuilder<T> begin(T pare
+        Preconditions.checkNotNull(renderType, "Render type must not be null");
+        Preconditions.checkNotNull(layers, "Layers must not be null");
+        Preconditions.checkArgument(layers.length > 0, "At least one layer must be specified");
+        Preconditions.checkArgument(Arrays.stream(layers).anyMatch(i -> i < 0), "All layers must be >= 0");
```
This should either be `allMatch(i -> i >= 0)` or `noneMatch(i -> i < 0)`

amadornes replied:
Fixed in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on src/main/java/net/minecraftforge/client/model/pipeline/RemappingVertexPipeline.java:
```diff
@@ -0,0 +1,116 @@
+    private final Vector3f normal = new Vector3f(0, 0, 0);
+    private final int[] color = new int[] { 255, 255, 255, 255 };
+    private final float[] uv0 = new float[] { 0, 0 };
+    private final int[] uv1 = new int[] { 0, 0 };
```
This should potentially default to `{ 0, 10 }`, otherwise remapping from a vertex format without the overlay element to a vertex with the overlay element will result in a red tint (which is the hurt overlay used for living entities) on the rendered quads.

amadornes replied:
The current version of the pipeline doesn't do it, but if that's what's considered the "default" state for that element, I can switch to it.

amadornes replied:
Fixed in [848f3ed](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/848f3ed9383583289102ff4859d2cf48b23b5360)

XFactHD commented on src/main/java/net/minecraftforge/client/textures/ITextureAtlasSpriteLoader.java:
```diff
@@ -8,13 +8,12 @@
 
 /**
- * A loader for custom TextureAtlasSprite implementations.<br>
+ * A loader for custom {@link TextureAtlasSprite texture atlas sprites}.<br>
```
`@linkplain` instead of `@link`

amadornes replied:
Fixed in [f48934a](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/f48934a8488de1c8fa613f4b21d72de611286c6e)

XFactHD commented on src/main/java/net/minecraftforge/common/util/TransformationHelper.java:
```diff
@@ -3,18 +3,16 @@
  */
 
-package net.minecraftforge.common.model;
+package net.minecraftforge.common.util;
```
Is there any point in this being a common class? I feel like this should be in the client package

amadornes replied:
`Transformation` is part of `com.mojang.math` and can be used for non-rendering-related stuff, so I just left it in common code - didn't really see a reason to move it

XFactHD replied:
Makes sense. I completely overlooked that those vanilla classes are not client-only

XFactHD reviewed (commented):
Took a rough look over the PR and left a few comments when something jumped out to me.

There is one thing I only commented on in one case which is an "issue" in a bunch of places: explicit synchronisation or the use of concurrent data structures in the startup events you added is unnecessary because they all fire sequentially and either fire on the main client thread or don't modify any data structures that may be accessed on the main client thread.

amadornes committed:
Additional event cleanup pass

As discussed on Discord:
- Remove "@hidden" and "@see <callsite>" javadoc annotations from all client events and replace them with @ApiStatus.Internal annotation
- Make all events that shouldn't be fired directly into abstract classes with protected constructors
- Another styling pass, just in case (caught some missed classes)

amadornes committed:
Add proper deprecation javadocs and de-dupe some vertex consumer code

LexManos commented:
Did a quick skim, and jesus this is a lot of unrelated/big changes.
Few things tho:
1) We need a tsrg of all renames/refactors so we can have them documented and can auto apply them using our existing tools.
2) Why do all of your extensions have of functions? Our extension system is designed to be implemented on the classes themselves thus not needing any casting/instanceofs...
3) The render layer system was a predicate to allow for faster checking. Can that system be moved to enumset/bitmasks instead of Sets?
4) All of your patched in init() functions, probably better to do a single callback so the patch is smaller, and have the init functions be package private?

amadornes commented:
Yep. It's a lot. I would've liked to split this into multiple PRs - first cleaning up, and then implementing all the big breaking changes separately, but due to how time-constrained we are and to avoid having to re-review the same files multiple times, we agreed to just do it all in one go.

As for your concerns:
1. Working on it! I already talked to Shrimp and I'll be providing the TSRG for all my changes myself instead of having to reference the list I provided in the renames issue
2. It's actually renaming existing classes (for example `IItemRenderProperties` -> `IClientItemExtensions`) and moving the lookup from an external class to the interface itself. They're essentially extensions to classes in common code, but with methods that only apply to client code (and in some cases return client-only types), so I figured renaming them to client extensions would make their purpose clearer
3. I was discussing this with @covers1624 earlier and we reckon the current predicate-based implementation may actually be slower than expected due to having to iterate through every chunk position for every render type just to fire the check, with all the duplicate `BlockState` lookups, solidity checks and other overhead that that involves. Turning it into a set means that we can just iterate through every chunk position, and for each of those just go through the model's supported render types.  
You make a good point that a mask-based system would be ideal, though, so I've gone ahead and implemented it using `BitSet` and a type-safe wrapper class. It could've been implemented up to vanilla's spec using just a `byte`, but my PR allowing for custom chunk render types would've had to revert it back to `BitSet`, so we may as well have it in from the start and avoid re-implementing all the logic
5. I considered making a monolithic initialization method, but since I had discussed with @OrionDevelopment that we wanted to keep hooks separate so they could be relocated if necessary, I figured I'd do it this way. If you want, I can switch it over to a single call to `ForgeHooksClient` though

amadornes committed:
Replace sets of chunk render types with a faster BitSet-backed collection

This largely addresses potential performance concerns that using a plain HashSet might involve by making lookups and iteration as linear as they can likely be (aside from using a plain byte/int/long for bit storage). Further performance concerns related to the implementation may be addressed separately, as all the implementation details are hidden from the end user

marchermans commented on src/main/java/net/minecraftforge/client/MinecraftForgeClient.java:
```diff
@@ -5,67 +5,41 @@
     public static Locale getLocale()
     {
         return Minecraft.getInstance().getLanguageManager().getSelected().getJavaLocale();
     }
```
Make into util class with static method.

amadornes replied:
Moved to `IForgeMinecraft` [848f3ed](https://github.com/MinecraftForge/MinecraftForge/pull/8786/commits/848f3ed9383583289102ff4859d2cf48b23b5360)

marchermans commented on src/main/java/net/minecraftforge/client/MinecraftForgeClient.java:
```diff
@@ -89,58 +63,4 @@ public static void releaseStencilBit(int bit)
             stencilBits.set(bit);
         }
     }
```
Move to a stencil manager instead of this unrelated class.

marchermans commented on src/main/java/net/minecraftforge/client/ChunkRenderTypeSet.java:
```diff
@@ -0,0 +1,208 @@
+ * An immutable ordered set (not implementing {@link java.util.Set}) of chunk {@linkplain RenderType render types}.
+ * <p>
+ * Considerably speeds up lookups and merges of sets of chunk {@linkplain RenderType render types}.
+ */
```
Please document that the user needs to intern this themselves since these are else very very short lived objects which can cause a decent amount of GC pressure.

marchermans commented on src/main/java/net/minecraftforge/client/model/generators/loaders/DynamicBucketModelBuilder.java:
```diff
@@ -73,16 +73,16 @@ public JsonObject toJson(JsonObject json)
         json.addProperty("fluid", fluid.toString());
```
Class is still named wrongly.

marchermans commented on src/main/java/net/minecraftforge/client/model/pipeline/QuadBakingVertexConsumer.java:
```diff
@@ -0,0 +1,144 @@
+ */
+public class QuadBakingVertexConsumer implements VertexConsumer
+{
+    private static final int STRIDE = IQuadTransformer.STRIDE;
```
Static imports.

marchermans reviewed (changes requested):
This is good as it stands, without respecting the comments, those still need to be fixed.

With regards to the refactor/review state. This PR reverts most of the VertexFormat aware rendering additions forge mades, and falls back to assuming that the pipeline is invoked with exactly the vanilla block format.

This does not hold in all cases, and breaks especially when shaders are in play (optifine and others) but it also breaks in vanilla cases where BERs are in use.

amadornes committed:
Requested changes

- Remove MinecraftForgeClient and move members to Minecraft, IForgeMinecraft and StencilManager
- Allow non-default elements to be passed into VertexConsumer and add support to derived classes
- Move array instantiation out of quad processing in lighting pipeline
- Fix flipped fluid container model
- Set default UV1 to the correct values in the remapping pipeline
- Minor documentation changes

XFactHD commented on src/main/java/net/minecraftforge/client/model/pipeline/RemappingVertexPipeline.java:
```diff
@@ -79,26 +103,39 @@ public VertexConsumer uv2(int u, int v)
+                continue;
+
+            // Try to match and output any of the supported elements, and if that fails, treat as misc
+            if (element.equals(DefaultVertexFormat.ELEMENT_POSITION))
```
Wouldn't a reference comparison be more appropriate here? The elements are constants anyway.

amadornes replied:
The vanilla ones are, but there's a chance someone might make an equivalent element that isn't the same object, and this should match those by usage/index/size as well.

marchermans reviewed (changes requested):
Codewise this is good.
To be able to test this I request the test mods to be properly commited as discussed.

Shadows-of-Fire commented:
@amadornes Please add "Fixes #8362" to the main description of this pr to link that issue (as changes made resolve it)

amadornes committed:
Add/update EXC entries and fix AT comment

amadornes committed:
Add test mod as per Orion's request

sciwhiz12 commented on patches/minecraft/net/minecraft/client/multiplayer/ClientLevel.java.patch:
```diff
@@ -102,5 +103,17 @@
++   public float getShade(float normalX, float normalY, float normalZ, boolean shade) {
++      boolean constantAmbientLight = this.m_104583_().m_108885_();
++      if (!shade)
++         return constantAmbientLight ? 0.9F : 1.0F;
```
Where are these constants derived from?

amadornes replied:
Vanilla's own `getShade(Direction side, boolean shade)` method in the same class. This is a more generic version based on normal vectors. I'll add a comment to the interface that registers the method explaining it.

amadornes replied:
Actually, my bad. The method is actually already documented in `IForgeBlockAndTintGetter`. I'll add the missing `@Override`, though.

sciwhiz12 commented on patches/minecraft/net/minecraft/client/renderer/RenderType.java.patch:
```diff
@@ -34,3 +34,20 @@
++
++   // FORGE START
++   private int chunkLayerId = -1;
++   public final int getChunkLayerId() {
```
Is this chunk layer ID only used for `ChunkRenderTypeSet`? Would mods have a reason to use this? Should this have a (javadoc) comment explaining briefly what this is and what values it returns?

amadornes replied:
There's no real use outside `ChunkRenderTypeSet`. I'll add a short javadoc.

sciwhiz12 commented on patches/minecraft/net/minecraft/client/renderer/block/model/ItemTransform.java.patch:
```diff
@@ -1,14 +1,38 @@
+    public final Vector3f f_111757_;
++   public final Vector3f rightRotation;
+ 
+    public ItemTransform(Vector3f p_111760_, Vector3f p_111761_, Vector3f p_111762_) {
```
Should this constructor be deprecated?

amadornes replied:
A second rotation is completely optional and most people have no need for it. Since the default value for it is pre-populated with a null rotation if the original constructor is used, I think it'd be fine to leave it as-is.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ChunkRenderTypeSet.java:
```diff
@@ -0,0 +1,210 @@
+    {
+        if (renderTypes.isEmpty())
+            return none();
+        if (renderTypes instanceof Set && renderTypes.size() == CHUNK_RENDER_TYPES.length)
+            return all();
```
What happens when the `renderTypes` contains a non-chunk render type and the size of the collection matches the known amount of chunk render types?

amadornes replied:
It... fails. Good point. I'll remove this check since it'd be kinda pointless to potentially iterate through everything twice to check.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/DimensionSpecialEffectsManager.java:
```diff
@@ -0,0 +1,60 @@
+ * <p>
+ * Provides a lookup by dimension type.
+ */
+public class DimensionSpecialEffectsManager
```
This is a utility class, so it should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/DimensionSpecialEffectsManager.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    /**
+     * Pre-registers vanilla dimension effects.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/DimensionSpecialEffectsManager.java:
```diff
@@ -0,0 +1,60 @@
+    private static DimensionSpecialEffects preRegisterVanillaEffects(Map<ResourceLocation, DimensionSpecialEffects> effects)
+    {
+        var overworldEffects = new DimensionSpecialEffects.OverworldEffects();
+        effects.put(BuiltinDimensionTypes.OVERWORLD_EFFECTS, overworldEffects);
+        effects.put(BuiltinDimensionTypes.NETHER_EFFECTS, new DimensionSpecialEffects.NetherEffects());
+        effects.put(BuiltinDimensionTypes.END_EFFECTS, new DimensionSpecialEffects.EndEffects());
```
Is there a place where these are defined in vanilla?

amadornes replied:
Yep. The `DimensionSpecialEffects` class has a map. I'll add a comment.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/EntitySpectatorShaderManager.java:
```diff
@@ -0,0 +1,46 @@
+ * <p>
+ * Provides a lookup.
+ */
+public class EntitySpectatorShaderManager
```
This is a utility class, so it should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/NamedRenderTypeManager.java:
```diff
@@ -0,0 +1,58 @@
+ * <p>
+ * Provides a lookup.
+ */
+public class NamedRenderTypeManager
```
This is a utility class, so it should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/RecipeBookManager.java:
```diff
@@ -0,0 +1,68 @@
+ * <p>
+ * Provides a recipe category lookup.
+ */
+public class RecipeBookManager
```
This is a utility class, so it should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/RecipeBookManager.java:
```diff
@@ -0,0 +1,68 @@
+    private static final Map<RecipeBookType, List<RecipeBookCategories>> TYPE_CATEGORIES = new HashMap<>();
+    private static final Map<RecipeType<?>, Function<Recipe<?>, RecipeBookCategories>> RECIPE_CATEGORY_LOOKUPS = new HashMap<>();
+
+    //Unmodifiable view, the nested lists are immutable
```
Formatting: space between comment header and comment body.
```suggestion
    // Unmodifiable view, the nested lists are immutable
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/StencilManager.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.BitSet;
+
+public class StencilManager
```
This is a utility class, so it should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/ModelEvent.java:
```diff
@@ -0,0 +1,139 @@
+        }
+
+        /**
+         * @return the model manager
```
This should use the inline `{@return}` tag.
```suggestion
         * {@return the model manager}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/ModelEvent.java:
```diff
@@ -0,0 +1,139 @@
+        }
+
+        /**
+         * @return the modifiable registry map of models and their model names
```
This should use the inline `{@return}` tag.
```suggestion
         * {@return the modifiable registry map of models and their model names}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/ModelEvent.java:
```diff
@@ -0,0 +1,139 @@
+        }
+
+        /**
+         * @return the model loader
```
This should use the inline `{@return}` tag.
```suggestion
         * {@return the model loader}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/RegisterEntitySpectatorShadersEvent.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Map;
+
+/**
+ * Allows users to register custom shaders to be used when the player spectates a certain kind of entity.
+ * Vanilla examples of this are the green effect for creepers and the invert effect for endermen.
```
What happens when an entity registers a shader for the player entity type, as is done in the `ShaderFixTest` test mod?

amadornes replied:
The same thing that happened with the previous registration method. It's just been moved to a standalone registration event instead of being a static register method in some class.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/RegisterGuiOverlaysEvent.java:
```diff
@@ -0,0 +1,119 @@
+        register(Ordering.AFTER, null, id, overlay);
+    }
+
+    // Synchronized to prevent multiple mods from registering at once
+    private synchronized void register(@NotNull Ordering ordering, @Nullable ResourceLocation other, @NotNull String id, @NotNull IGuiOverlay overlay)
```
Is this event fired in parallel?

amadornes replied:
It's not! I missed that one when I removed `synchronized` from another method. I originally intended for registration events to fire in parallel, but they can't due the way things work with the event bus right now.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/RegisterNamedRenderTypesEvent.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * Registers a named render type.<p/>
```
Self-closing paragraph tags are invalid. (Same for the other occurences in this class.)

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/RegisterRecipeBookCategoriesEvent.java:
```diff
@@ -0,0 +1,72 @@
+     */
+    public void registerAggregateCategory(RecipeBookCategories category, List<RecipeBookCategories> others)
+    {
+        aggregateCategories.put(category, others instanceof ImmutableList il ? il : ImmutableList.copyOf(others));
+    }
+
+    /**
+     * Registers the list of categories that compose a recipe book.
+     */
+    public void registerBookCategories(RecipeBookType type, List<RecipeBookCategories> categories)
+    {
+        typeCategories.put(type, categories instanceof ImmutableList il ? il : ImmutableList.copyOf(categories));
```
The ternary to check if the instance is already an `ImmutableList` is unneeded, as [`ImmutableList.copyOf(Collection)`](https://guava.dev/releases/31.0-jre/api/docs/com/google/common/collect/ImmutableList.html#copyOf(java.util.Collection)) is already noted to avoiding copying the collection when it is safe to do so. 

Even though the javadoc says that the conditions are undocumented and subject to change, the implementation suggests that it is done when the collection is a `ImmutableCollection` and not a partial view of a list, which is likely to be any `ImmutableList` passed to this method.

amadornes replied:
I believe this was a straight code port from the original recipe book PR, but you're right. I'll change it.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/ScreenEvent.java:
```diff
@@ -290,50 +273,70 @@ public PoseStack getPoseStack()
-     *   <li>{@link Result#DENY} - forcibly renders the potion indicators in <em>classic</em> mode.</li>
-     * </ul>
+     * <p>This event is {@linkplain Cancelable cancellable} and does not {@linkplain HasResult have a result}.</p>
+     * Cancelling this event will prevent vanilla rendering.
```
The closing paragraph tag should be after the sentence describing what cancelling the event does.
```suggestion
     * <p>This event is {@linkplain Cancelable cancellable} and does not {@linkplain HasResult have a result}.
     * Cancelling this event will prevent vanilla rendering.</p>
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/event/ScreenEvent.java:
```diff
@@ -987,25 +933,94 @@ public Pre(Screen screen, char codePoint, int modifiers)
+        }
+
+        /**
+         * Gets the currently open screen at the time of the event being fired.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/extensions/common/IClientBlockExtensions.java:
```diff
@@ -14,15 +14,32 @@
+import java.util.function.Consumer;
+
+/**
+ * {@link LogicalSide#CLIENT Client-only} extensions to {@link Block}.
```
The label should be plain text, so the tag should be `{@linkplain}`.
```suggestion
 * {@linkplain LogicalSide#CLIENT Client-only} extensions to {@link Block}.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/extensions/common/IClientFluidTypeExtensions.java:
```diff
@@ -15,41 +15,47 @@
 /**
- * An interface which defines properties relating to how a {@link FluidType}
- * should render.
+ * {@link LogicalSide#CLIENT Client-only} extensions to {@link FluidType}.
```
The label should be plain text, so the tag should be `{@linkplain}`.
```suggestion
 * {@linkplain LogicalSide#CLIENT Client-only} extensions to {@link FluidType}.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/extensions/common/IClientItemExtensions.java:
```diff
@@ -0,0 +1,143 @@
+import java.util.function.Consumer;
+
+/**
+ * {@link LogicalSide#CLIENT Client-only} extensions to {@link Item}.
```
The label should be plain text, so the tag should be `{@linkplain}`.
```suggestion
 * {@linkplain LogicalSide#CLIENT Client-only} extensions to {@link Item}.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/extensions/common/IClientMobEffectExtensions.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.function.Consumer;
+
+/**
+ * {@link LogicalSide#CLIENT Client-only} extensions to {@link MobEffect}.
```
The label should be plain text, so the tag should be `{@linkplain}`.
```suggestion
 * {@linkplain LogicalSide#CLIENT Client-only} extensions to {@link MobEffect}.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/gui/ClientTooltipComponentManager.java:
```diff
@@ -0,0 +1,47 @@
+ * <p>
+ * Provides a lookup.
+ */
+public class ClientTooltipComponentManager
```
This is a utility class, so it should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/gui/overlay/ForgeGui.java:
```diff
@@ -3,72 +3,77 @@
+        debugOverlay = new ForgeDebugScreenOverlay(mc);
+    }
+
+    public final Minecraft getMinecraft()
```
Why is this `final`?

amadornes replied:
No real reason. Good point.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/gui/overlay/GuiOverlayManager.java:
```diff
@@ -0,0 +1,75 @@
+ * <p>
+ * Provides a lookup by ID, as well as all registered {@link IGuiOverlay overlays}.
+ */
+public class GuiOverlayManager
```
This is a utility class, so it should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/gui/overlay/NamedGuiOverlay.java:
```diff
@@ -0,0 +1,18 @@
+import net.minecraft.resources.ResourceLocation;
+
+/**
+ * An object representation of an {@link IGuiOverlay overlay} with a name.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/gui/overlay/NamedGuiOverlay.java:
```diff
@@ -0,0 +1,18 @@
+ * An object representation of an {@link IGuiOverlay overlay} with a name.<p/>
+ * Useful to identify overlays in {@link net.minecraftforge.client.event.RenderGuiOverlayEvent}.
+ * <p>
+ * Users should not be instantiating this themselves. Retrieve from {@link GuiOverlayManager}.
```
Perhaps the canonical constructor should be `@Deprecated` then?

amadornes replied:
I'll mark it as `@ApiStatus.Internal`, since that's what we generally seem to be going for now.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/gui/overlay/VanillaGuiOverlay.java:
```diff
@@ -0,0 +1,224 @@
+import org.lwjgl.opengl.GL11;
+
+/**
+ * All the vanilla {@link IGuiOverlay HUD overlays} in the order that they render.
```
The label should be in plain text, so the tag should be `{@linkplain}`.
```suggestion
 * All the vanilla {@linkplain IGuiOverlay HUD overlays} in the order that they render.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/CompositeModel.java:
```diff
@@ -5,366 +5,428 @@
+import java.util.function.Function;
+
+/**
+ * A model composed of several named children.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/DynamicFluidContainerModel.java:
```diff
@@ -0,0 +1,303 @@
+
+/**
+ * A dynamic fluid container model, capable of re-texturing itself at runtime to match the contained fluid.<p/>
+ * Composed of a base layer, a fluid layer (applied with a mask) and a cover layer (optionally applied with a mask).<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/DynamicFluidContainerModel.java:
```diff
@@ -0,0 +1,303 @@
+ *
+ * @see Colors
+ */
+public final class DynamicFluidContainerModel implements IUnbakedGeometry<DynamicFluidContainerModel>
```
Why is this `final`?

amadornes replied:
I believe it already was, so I left it as-is. Will change.

sciwhiz12 commented on src/main/java/net/minecraftforge/common/util/ConcatenatedListView.java:
```diff
@@ -0,0 +1,148 @@
+import java.util.function.Supplier;
+
+/**
+ * A list that concatenates multiple other lists for efficient iteration.<p/>
+ * You may use this in place of creating a new list and calling {@link List#addAll(Collection)}
+ * for each of your collections.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/common/util/ConcatenatedListView.java:
```diff
@@ -0,0 +1,148 @@
+    @Override public boolean retainAll(@NotNull Collection<?> c) { throw new UnsupportedOperationException(); }
+    @Override public void clear() { throw new UnsupportedOperationException(); }
+
+    // Other unsupported operations - we could support these, but effort
+    @NotNull @Override public ListIterator<T> listIterator() { throw new UnsupportedOperationException(); }
+    @NotNull @Override public ListIterator<T> listIterator(int index) { throw new UnsupportedOperationException(); }
+    @NotNull @Override public List<T> subList(int fromIndex, int toIndex) { throw new UnsupportedOperationException(); }
```
This means this class does not fully implement the contract of `List`, and should be noted as such in its javadocs and possibly in any place which uses this class.

amadornes replied:
I'll give them a default implementation falling back to `ArrayList` since I have a concatenation method in place already.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/textures/TextureAtlasSpriteLoaderManager.java:
```diff
@@ -0,0 +1,44 @@
+ * <p>
+ * Provides a lookup.
+ */
+public class TextureAtlasSpriteLoaderManager
```
This is a utility class and should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/textures/ITextureAtlasSpriteLoader.java:
```diff
@@ -8,13 +8,12 @@
 
 /**
- * A loader for custom TextureAtlasSprite implementations.<br>
+ * A loader for custom {@linkplain  TextureAtlasSprite texture atlas sprites}.<br>
```
There is an additional space in the tag.
```suggestion
 * A loader for custom {@linkplain TextureAtlasSprite texture atlas sprites}.<br>
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/renderable/BakedModelRenderable.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.Arrays;
+
+/**
+ * {@link IRenderable Renderable } wrapper for {@link BakedModel baked models}.<p/>
```
There is an additional space in the `{@link}` tag. Additionally, self-closing paragraph tags are invalid. Additionally, the label should be in plain text, so the tag should be `{@linkplain}`.
```suggestion
 * {@link IRenderable Renderable} wrapper for {@linkplain BakedModel baked models}.<p>
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/pipeline/QuadBakingVertexConsumer.java:
```diff
@@ -0,0 +1,164 @@
+import static net.minecraftforge.client.model.IQuadTransformer.*;
+
+/**
+ * Vertex consumer that outputs {@link BakedQuad baked quads}.
```
The label should be plain text, so the tag should be `{@linkplain}`.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/obj/ObjTokenizer.java:
```diff
@@ -7,24 +7,27 @@
 
-public class LineReader implements AutoCloseable
+/**
+ * A tokenizer for OBJ and MTL files.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/EmptyModel.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.function.Function;
+
+/**
+ * A completely empty model with no quads or texture dependencies.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/IModelBuilder.java:
```diff
@@ -6,50 +6,110 @@
+import java.util.List;
+
+/**
+ * Base interface for any object that may collect culled and unculled faces and bake them into a model.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/ItemLayerModel.java:
```diff
@@ -5,488 +5,170 @@
+public class ItemLayerModel implements IUnbakedGeometry<ItemLayerModel>
 {
-    public static final ItemLayerModel INSTANCE = new ItemLayerModel(ImmutableList.of());
+    private static final Logger LOGGER = LogManager.getLogger();
```
This should use `LogUtil.getLogger()` and SLF4J's `Logger`.

amadornes replied:
As discussed, since loggers in this PR were borrowed from other Forge code and seems like a lot of places are still using the Log4J API instead of SLF4J, I'll keep it as-is and it can be updated in a future cleanup pass if Forge decides to fully transition to native SLF4J.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/geometry/BlockGeometryBakingContext.java:
```diff
@@ -3,48 +3,51 @@
+import java.util.function.Function;
+
+/**
+ * A {@link IGeometryBakingContext geometry baking context} that is bound to a {@link BlockModel}.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/geometry/GeometryLoaderManager.java:
```diff
@@ -0,0 +1,57 @@
+ * <p>
+ * Provides a lookup.
+ */
+public class GeometryLoaderManager
```
This is a utility class, and should be `final` with a `private` constructor.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/geometry/IGeometryBakingContext.java:
```diff
@@ -0,0 +1,93 @@
+
+/**
+ * The context in which a geometry is being baked, providing information such as lighting and
+ * {@link ItemTransforms transforms}, and allowing the user to create {@link Material materials} and query
+ * {@link RenderTypeGroup render types}.
```
These labels should be in plain text, and therefore should be `{@linkplain}`.
```suggestion
 * {@linkplain ItemTransforms transforms}, and allowing the user to create {@linkplain Material materials} and query
 * {@linkplain RenderTypeGroup render types}.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/geometry/IGeometryBakingContext.java:
```diff
@@ -0,0 +1,93 @@
+public interface IGeometryBakingContext
+{
+    /**
+     * Gets the name of the model being baked for logging and caching purposes.
```
The javadocs for these getters should use the inline `{@return}` tag.
```suggestion
     * {@return the name of the model being baked for logging and caching purposes}
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/geometry/IGeometryLoader.java:
```diff
@@ -0,0 +1,26 @@
+import net.minecraftforge.client.event.RegisterClientReloadListenersEvent;
+
+/**
+ * A loader for custom {@link IUnbakedGeometry model geometries}.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/geometry/SimpleUnbakedGeometry.java:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * Base class for implementations of {@link IUnbakedGeometry} which do not wish to handle model creation themselves,
+ * instead supplying {@link BakedQuad baked quads} through a builder.
```
This label should be in plain text, and therefore use `{@linkplain}`.
```suggestion
 * instead supplying {@linkplain BakedQuad baked quads} through a builder.
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/lighting/QuadLighter.java:
```diff
@@ -0,0 +1,152 @@
+import static net.minecraftforge.client.model.IQuadTransformer.*;
+
+/**
+ * Base class for all quad lighting providers.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/model/obj/ObjLoader.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.Map;
+
+/**
+ * A loader for {@link ObjModel OBJ models}.<p/>
```
Self-closing paragraph tags are invalid.

sciwhiz12 reviewed (changes requested):
I've not made a comprehensive review of everything here, as I believe @OrionDevelopment and @covers1624 are doing that with your input in a voice call. Most of my comments are due to the self-closing paragraph tags sprinkled throughout the javadoc comments and other formatting details, and I've likely missed a few tags here and there.

amadornes committed:
Additional requested changes

sciwhiz12 reviewed (approved):
:shipit:

amadornes committed:
Allow custom model types to request the particle texture to be loaded

marchermans reviewed (approved):
Final testing passed.
This is good to go.

:shipit:

amadornes committed:
Even more requested changes

marchermans reviewed (approved):
Thanks

LexManos commented:
So, this has been vetted and approved by the crew. However due to its size, and the amount of things it breaks, we have decided to hold off on this until 1.19.1's release. And roll it in/quickly with the MC version update. Along with our pass of renames.

We are still waiting on Ama to provide his renaming/updating guide. So please do that ASAP.

XFactHD committed:
Revert "Allow safely registering RenderType predicates at any time (#8685)"

This reverts commit be7275443fd939db9c58bcad47079c3767789ac1.

amadornes committed:
Improve generics in ConcatenatedListView and add missing fallbacks

amadornes commented:
As originally discussed in #8685, and following up additional chats on Discord, that PR has been reverted as it is superseded by the new handling of render types via baked models. Mods that need to make changes to a block's render types after game startup should now be doing so direcly via the model. This force-push also fixes additional merge conflicts.

With these changes and the next commit fixing an oversight that didn't get previously caught, this PR is ready to be merged.

Let me know before that happens so I can release the latest version of the update guide.

amadornes committed:
Fix fluid render types being bound to the fluid and not its holder

amadornes committed:
Remove non-contractual nullability in ChunkRenderTypeSet and add isEmpty

Additionally, introduce chunk render type checks in ItemBlockRenderTypes

cpw commented:
Yeet

amadornes commented:
Oh god I'll go over the update guide one last time and post it!
